README.md

```diff
@@ -514,6 +514,86 @@ $ grpcurl -plaintext 127.0.0.1:9093 spacemesh.v1.DebugService.NetworkInfo
+
+For an existing database, the `PRAGMA user_version` is checked against the expected version number. If the database's schema version is too new, go-spacemesh fails right away as an older go-spacemesh version cannot be expected to work with a database from a newer version. If the database version number is older than the expected version, go-spacemesh runs the necessary migration scripts embedded in go-spacemesh executable and updates `PRAGMA user_version = ...`. The migration scripts are located in the following folders:
+* `sql/statesql/schema/migrations` for `state.sql`
+* `sql/localsql/schema/migrations` for `local.sql`
+
+Additionally, some migrations ("coded migrations") can be implemented in Go code, in which case they reside in `.go` files located in `sql/statesql` and `sql/localsql` packages, respectively. It is worth noting that old coded migrations can be removed at some point, rendering database versions that are *too* old unusable with newer go-spacemesh versions.
+
+After all the migrations are run, go-spacemesh compares the schema of each database to the embedded schema scripts and if they differ, warns the user about any differences:
+```
+logger.go:146: 2024-06-05T05:39:32.247+0400	WARN	database schema drift detected	{"uri": "file:/var/folders/r0/4mks2v4n5ysbntnf3xq6h_q80000gn/T/TestSchemaidempotent_migration3425594786/001/test.db", "diff": "  (\n  \t\"\"\"\n  \t... // 81 identical lines\n  \t    PRIMARY KEY (kind, epoch)\n  \t) WITHOUT ROWID;\n- \t\n- \t-- some change\n  \t\"\"\"\n  )\n"}
+```
+
+In this case, an empty line and `-- some change` was added to `schema.sql` by hand. The pretty-printed diff looks like this:
+```
+  (
+  	"""
+  	... // 81 identical lines
+  	    PRIMARY KEY (kind, epoch)
+  	) WITHOUT ROWID;
+-
+- 	-- some change
+  	"""
+  )
+```
+
+The possible reasons for schema drift can be the following:
+* running an unreleased version of go-spacemesh using your data folder. The unreleased version may contain migrations that may be changed before the release happens
+* manual changes in the database
+* external SQLite tooling used on the database that adds some tables, indices etc.
+
+In the latter case, it is possible to make go-spacemesh ignore certain objects (tables and indices) when checking for schema drift. For this, you can use `main.db-schema-ignore-rx` setting to set a regular expression that is used to ignore tables and indices in the database during schema drift checks. The setting defaults to `_litestream` to help with certain tooling.
+
+The schema changes in go-spacemesh code should be always done by means of adding migrations. After that, the schema tests in `sql/localsql` and `sql/statesql` will start failing. When the tests fail, they display the difference between the schema stored in `schema.sql` and the schema that is loaded from the database after running all the migrations.
+If the schema changes shown in the diff are expected, the schema file needs to be updated.
```

Comment:
The `README.md` is primarily written for users - and this is already TMI for users. If the average user runs into a schema drift it's very unlikely they will be able to resolve that themselves.

Reply:
Moved these docs to `CONTRIBUTING.md`